src/RandomSetGenerator.java

```diff
@@ -0,0 +1,106 @@
+
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+
+public class RandomSetGenerator {
+
+	static private Random randomNumberGenerator;
+	
+	public RandomSetGenerator() {
+		this(new SecureRandom()); 
+	}
+
+	public RandomSetGenerator(Random generator) {
+		randomNumberGenerator = generator;
+	}
+
+	public List<Integer> generate(int quantity, int maxValue) throws IllegalArgumentException {
+		Set<Integer> randomSet;
+		if(quantity <= 0 || maxValue <= 0 || quantity > maxValue) {
+			throw new IllegalArgumentException("invalid inputs to generate(). Inputs: quantity: " + quantity + ", maxValue: " + maxValue);
+		}
+		
+		// Random.nextInt(maxvalue) is exclusive
+		maxValue++;
+		
+		if( quantity < maxValue / 2) {
+			randomSet = new HashSet<Integer>(quantity);
+			
+			for(int i = 0; i < quantity; i++) {
+				// get random int, range 0 to maxValue + 1 because Random.nextInt(n) is exclusive
+				Integer nextInt;
+				do {
+					nextInt = randomNumberGenerator.nextInt(maxValue);
+				} while (randomSet.contains(nextInt));
+				randomSet.add(nextInt);
+			}
+			return new LinkedList<Integer>(randomSet);
+		} else {
+			quantity = maxValue - quantity;
+			randomSet = new HashSet<Integer>(quantity);
+			for(int i = 0; i < quantity; i++) {
+				// get random int, range 0 to maxValue + 1 because Random.nextInt(n) is exclusive
+				Integer nextInt;
+				do {
+					nextInt = randomNumberGenerator.nextInt(maxValue);
+				} while (randomSet.contains(nextInt));
+				randomSet.add(nextInt);
+			}
+			List<Integer> list = new LinkedList<Integer>();
+			for(int i = 0; i < maxValue; i++) {
+				list.add(i);
```

Comment:
instead of calling removeAll() after this, just check to see if the set contains 'i' and don't add the number. that way, we only make one pass from 0 to maxValue.